jena-arq/src/main/java/org/apache/jena/sparql/expr/aggregate/AggMedian.java

```diff
@@ -0,0 +1,118 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.sparql.expr.aggregate;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+
+import org.apache.commons.math3.stat.descriptive.rank.Median;
+import org.apache.jena.graph.Node ;
+import org.apache.jena.sparql.engine.binding.Binding ;
+import org.apache.jena.sparql.expr.Expr ;
+import org.apache.jena.sparql.expr.ExprEvalException ;
+import org.apache.jena.sparql.expr.ExprList ;
+import org.apache.jena.sparql.expr.NodeValue ;
+import org.apache.jena.sparql.expr.nodevalue.XSDFuncOp ;
+import org.apache.jena.sparql.function.FunctionEnv ;
+
+public class AggMedian extends AggregatorBase
+{
+    // ---- MEDIAN(?var)
+    public AggMedian(Expr expr) { super("MEDIAN", false, expr) ; } 
+    @Override
+    public Aggregator copy(ExprList expr) { return new AggMedian(expr.get(0)) ; }
+
+    // XQuery/XPath Functions&Operators suggests zero
+    // SQL suggests null.
+    private static final NodeValue noValuesToMedian = NodeValue.nvZERO ; // null 
+
+    @Override
+    public Accumulator createAccumulator()
+    { 
+        return new AccMedian(getExpr()) ;
+    }
+
+    @Override
+    public Node getValueEmpty()     { return NodeValue.toNode(noValuesToMedian) ; } 
+    
+    @Override
+    public int hashCode()   { return HC_AggMedian ^ getExprList().hashCode() ; }
+
+    @Override
+    public boolean equals(Aggregator other, boolean bySyntax) {
+        if ( other == null ) return false ;
+        if ( this == other ) return true ;
+        if ( ! ( other instanceof AggMedian ) ) return false ;
+        AggMedian a = (AggMedian)other ;
+        return exprList.equals(a.exprList, bySyntax) ;
+    }
+    
+    // ---- Accumulator
+    private static class AccMedian extends AccumulatorExpr
+    {
+        // Non-empty case but still can be nothing because the expression may be undefined.
+        private NodeValue total = noValuesToMedian ;
+        private int count = 0 ;
+        ArrayList<NodeValue> collection=new ArrayList<NodeValue>(); 
+        
+        static final boolean DEBUG = false ;
+        
+        public AccMedian(Expr expr) { super(expr, false) ; }
+
+        @Override
+        protected void accumulate(NodeValue nv, Binding binding, FunctionEnv functionEnv)
+        { 
+            if ( DEBUG ) System.out.println("median: "+nv) ;
+
+            if ( nv.isNumber() )
+            {
+              count++ ;
+              collection.add(nv);
+            }
+            else
+            {
+                //ARQ.getExecLogger().warn("Evaluation error: median() on "+nv) ;
+                throw new ExprEvalException("median: not a number: "+nv) ;
+            }
+
+            if ( DEBUG ) System.out.println("median: ("+count+")") ;
```

Comment:
This should use a proper method from the logging framework, not `System.out.println`.

Reply:
Just to be clear, you can use SLF4J to log at this point.

Reply:
since this is also the case for Avg, Max and Min should that be changed as well?

Reply:
You needn't (not your code) but if you have time, that would be nice. I am already looking at this so it's not more review.

Reply:
I've done the change for median. if you approve this change I will do the same for avg, max and min as well.